internal/controllers/topology/machineset/machineset_controller.go

```diff
@@ -119,12 +120,26 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		return ctrl.Result{}, nil
 	}
 
+	// Create a patch helper to add or remove the finalizer can be garbage collected by Kubernetes.
+	patchHelper, err := patch.NewHelper(ms, r.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create patch helper for %s", tlog.KObj{Obj: ms})
+	}
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, ms); err != nil {
```

Comment:
Using the generic patch flow here, even though this controller should only ever try to patch metadata - adding or removing the finalizer.